server/src/main/java/io/druid/metadata/SQLMetadataSegmentManager.java

```diff
@@ -144,26 +142,35 @@ public void run()
           delay.getMillis(),
           TimeUnit.MILLISECONDS
       );
-      started = true;
+      lifecycleLock.started();
+    }
+    finally {
+      lifecycleLock.exitStart();
     }
   }
 
   @Override
   @LifecycleStop
   public void stop()
   {
-    synchronized (lock) {
-      if (!started) {
-        return;
-      }
+    if (!lifecycleLock.canStop()) {
+      return;
+    }
+    try {
+      final ConcurrentHashMap<String, DruidDataSource> emptyMap = new ConcurrentHashMap<>();
+      ConcurrentHashMap<String, DruidDataSource> current;
+      do {
+        current = dataSourcesRef.get();
```

Comment:
`current=dataSourcesRef.get()` could be called as soon as the `poll`  function is entered?here, there is no guarantee that `dataSourcesRef`  is safe. it may be updated in `for loop` above. @jihoonson @gianm @jon-wei

Reply:
This class was refactored extensively in #6370 and then renamed in #7306 to SqlSegmentsMetadataManager. So this issue might not still exist. Have you been able to take a look at the new code?

Reply:
Which for loop are you referring to? Can you link here? Also, there has been a pretty big refactoring for this class. Do you still see the same issue?